File: python/pipemaze.py
from dataclasses import dataclass

@dataclass(frozen=True)
class Point:
    x: int
    y: int

    def __add__(self, pt):
        return Point(self.x+pt.x, self.y+pt.y)

Point.Direction = {'N':Point(0,-1), 'S':Point(0,1), 'W':Point(-1,0), 'E':Point(1,0)}


class Maze:

    def __init__(self, input):
        self.maze = []
        self._start = None
        self.pipes = set()
        for y,line in enumerate(input.readlines()):
            row = [*line.strip()]
            if 'S' in row:
                self._start = Point(row.index('S'), y)
                self.pipes.add(self._start)
            self.maze.append(row)

    def get(self, pt):
        if pt.y < len(self.maze) and pt.x < len(self.maze[pt.y]):
            return self.maze[pt.y][pt.x]
        raise IndexError(f"{pt} not in maze")

    def trace(self):
        count = 1
        dir = self.start()
        at = self._start + Point.Direction[dir]
        while at != self._start:
            self.pipes.add(at)
            match self.get(at):
                case '|' | '-':
                    pass
                case 'L' | '7':
                    dir = {'N':'W', 'S':'E', 'E':'S', 'W':'N'}[dir]
                case 'J' | 'F':
                    dir = {'N':'E', 'S':'W', 'E':'N', 'W':'S'}[dir]
            at = at + Point.Direction[dir]
            count += 1
        return count

    def start(self):
        what = [d for d in Point.Direction
                  if (d,self.get(self._start+Point.Direction[d])) in [
                      ('N','|'),('S','|'),('W','-'),('E','-'),
                      ('N','7'),('E','7'),('W','L'),('S','L'),
                      ('N','F'),('W','F'),('E','J'),('S','J')]]
        match what:
            case ['N','S'] | ['S','N']:
                self.maze[self._start.y][self._start.x] = '|'
            case ['E','W'] | ['W','E']:
                self.maze[self._start.y][self._start.x] = '-'
            case ['E','N'] | ['N','E']:
                self.maze[self._start.y][self._start.x] = 'L'
            case ['E','S'] | ['S','E']:
                self.maze[self._start.y][self._start.x] = 'F'
            case ['N','W'] | ['W','N']:
                self.maze[self._start.y][self._start.x] = 'J'
            case ['S','W'] | ['W','S']:
                self.maze[self._start.y][self._start.x] = '7'
        return what[-1]

    def each(self):
        for y in range(len(self.maze)):
            for x in range(len(self.maze[y])):
                yield Point(x,y)

    def inside(self, start):
        count = 0
        row = self.maze[start.y]
        pipe = None
        x = start.x
        cols = range(x+1, len(row)) if len(row)-x < x else range(x)
        span =''.join(self.get(Point(i,start.y)) for i in cols if Point(i, start.y) in self.pipes)
        span = span.replace('-','').replace('L7','|').replace('FJ','|')
        return span.count('|')%2 == 1


def pipemaze1(maze):
    return maze.trace() // 2

def pipemaze2(maze):
    return sum(maze.inside(pt) for pt in maze.each() if pt not in maze.pipes)
    count = 0
    row = []
    for c in (maze.get(pt) if pt in maze.pipes else ('+' if maze.inside(pt) else ' ') for pt in maze.each()):
        row.append(c)
        if len(row) >= len(maze.maze[0]):
            count += row.count('+')
            print(''.join(row))
            row = []
    return count

def main():
    import sys
    maze = Maze(sys.stdin)
    print(pipemaze1(maze))
    print(pipemaze2(maze))

if __name__=='__main__':
    main()

File: lua/pipemaze.lua
local aoc = require"adventofcode"

-- Directions of travel
Direction = {
    E = {1, 0},
    W = {-1, 0},
    N = {0, -1},
    S = {0, 1}
}

-- Transform direction of travel
Transform = {
    ['|'] = {S='S', N='N',
    sides = {S={O={{1,0}}, I={{-1,0}} },
             N={O={{-1,0}}, I={{1,0}} }
            }
    },
    ['-'] = {E='E', W='W',
    sides = {W={O={{0,1}}, I={{0,-1}} },
             E={O={{0,-1}}, I={{0,1}} }
            }
    },
    ['L'] = {S='E', W='N',
    sides = {S={O={}, I={{-1,0},{-1,1},{0,1}} },
             W={O={{-1,0},{-1,1},{0,1}}, I={} }
            }
    },
    ['J'] = {S='W', E='N',
    sides = {S={O={{1,0},{1,1},{0,1}}, I={} },
             E={O={}, I={{1,0},{1,1},{0,1}} }
            }
    },
    ['7'] = {N='W', E='S',
    sides = {N={O={}, I={{1,0},{1,-1},{0,-1}} },
             E={O={{1,0},{1,-1},{0,-1}}, I={} }
            }
    },
    ['F'] = {W='S', N='E',
    sides = {N={O={{-1,0},{-1,-1},{0,-1}}, I={} },
             W={O={}, I={{-1,0},{-1,-1},{0,-1}} }
            }
    }
}

-- Mark coverage
--[[
 .0. .~. .0. .0. .~~ ~~.
 ~|+ 0-0 ~L0 0J~ 07~ ~F0
 .0. .+. ~~. .~~ .0. .0.
]]

Map = {}
Coverage = {}

function read_map(input)
    local start
    local y = 1
    for line in input:lines() do
        Map[#Map+1] = aoc.findall(line, ".")
        local S = line:find('S')
        if S then
            start = {S, y}
        end
        Coverage[#Coverage+1] = {}
        y = y + 1
    end
    return start
end

function cover(at, what)
    local row = Coverage[at[2]]
    if row and at[1] <= #Map[at[2]] then
        if what and row[at[1]] ~= 'P' then
            row[at[1]] = what
        end
        return row[at[1]]
    end
    return nil
end

function get(at)
    local row = Map[at[2]]
    return row and row[at[1]]
end

function move(from, to)
    return {from[1]+to[1], from[2]+to[2]}
end

function mark(at, dir)
    cover(at, 'P') -- mark the pipe itself
    local pipe = get(at)
    local sides
    if pipe == 'S' then
        if dir == 'N' or dir == 'S' then
            sides = Transform['|'].sides[dir]
        else -- dir == 'W' or dir == 'E'
            sides = Transform['-'].sides[dir]
        end
    else
        sides = Transform[pipe].sides[dir]
    end
    for i = 1,#sides.I do
        cover(move(at, sides.I[i]), 'I')
    end
    for i = 1,#sides.O do
        cover(move(at, sides.O[i]), 'O')
    end
end

function pick_a_direction(start)
    for dir,step in pairs(Direction) do
        local pipe = get(move(start, step))
        if Transform[pipe] and Transform[pipe][dir] then
            return dir
        end
    end
    error"Can't go anywhere!"
end

function inside(start)
    if cover(start) == 'P' then
        return false
    end
    local row = Map[start[2]]
    local function _inside(at,dir)
        local count = false
        local pipe = get(at)
        while pipe do
            if pipe == 'S' then
                dir[1] = -dir[1]
                return _inside(move(start,dir), dir)
            end
            if cover(at) == 'P' then
                if pipe == '|' or pipe == 'F' or pipe == '7' then
                    count = not count
                end
            end
            at = move(at, dir)
            pipe = get(at)
        end
        return count
    end
    local dir
    if start[1] <= #row-start[1] then
        dir = {-1,0}
    else
        dir = {1,0}
    end
    return _inside(move(start,dir), dir)
end

local BORDER = {{-1,-1},{0,-1},{1,-1},{-1,0},{1,0},{-1,1},{0,1},{1,1}}
function fill(from)
    local what = cover(from)
    local atlist = {from}
    while #atlist > 0 do
        local at = atlist[#atlist]
        atlist[#atlist] = nil
        --print(at[1],at[2])
        for _,d in ipairs(BORDER) do
            local p = move(at,d)
            if not cover(p) then
                cover(p, what)
                atlist[#atlist+1] = p
            end
        end
    end
end

function step(start, dir, count)
    start = move(start, Direction[dir])
    mark(start, dir)
    local pipe = get(start)
    if pipe == 'S' then
        return count
    end
    return step(start, Transform[pipe][dir], count+1)
end

function pipemaze1(start)
    local dir = pick_a_direction(start)
    -- I don't believe it's possible for this to be odd
    -- but just in case
    mark(start, dir)
    return math.ceil(step(start, dir, 1) / 2)
end

function pipemaze2(start)
    pipemaze1(start)
    local inmark, top
    for y = 1,#Map do
        for x = 1,#Map[y] do
            if cover{x,y} == 'I' or cover{x,y} == 'O' then
                if inside{x,y} then
                    inmark = cover{x,y}
                    top = y
                    goto foundmark
                end
            end
        end
    end
    error"No marks found"
    ::foundmark::
    for y = top,#Map do
        for x = 1,#Map[y] do
            if cover{x,y} == inmark then
                fill({x,y})
            end
        end
    end
    local count = 0
    for y = 1,#Map do
      for x = 1,#Map[y] do
        if Coverage[y][x] == inmark then
            count = count + 1
        end
      end
    end
    return count
end

-- main 
do
    local start = read_map(io.stdin)
    print(pipemaze1(start))
    print(pipemaze2(start))
end
